### PR TITLE
Add support for max_tokens = -1 for LlamaCPP

### PR DESCRIPTION
**Pull request by masands**
# Calculate max_tokens

Added a '_get_max_tokens' function to the LlamaCpp wrapper to calculate the max_tokens if the max_token paramter is set to -1 (similar to the BaseOpenAI wrapper). Setting max_tokens = -1 returns as many tokens as possible given the prompt and the models maximal context size.

## Who can review?
@agola11

### CODE REVIEW

File: langchain/llms/llamacpp.py [hunk 1]
```diff
@@ -282,3 +287,30 @@ def stream(
...
+        Args:
+            prompt: The prompt to use for generation.
+
+        Returns:
+            The maximum number of tokens to generate.
+        """
+        from llama_cpp import llama_cpp
+
+        prompt_tokens: List[llama_cpp.llama_token] = self.client.tokenize(
+            b" " + prompt.encode("utf-8")
```
Comment (dev2049) [inline langchain/llms/llamacpp.py:303]: why the extra space?

File: langchain/llms/llamacpp.py [hunk 2]
```diff
@@ -230,6 +230,11 @@ def _call(
             return combined_text_output
         else:
             params = self._get_parameters(stop)
+
+            """Set max tokens."""
```
Comment (dev2049) [inline langchain/llms/llamacpp.py:234]: could we comment with # (or remove comment)

File: langchain/llms/llamacpp.py [hunk 3]
```diff
@@ -282,3 +287,30 @@ def stream(
                     token=token, verbose=self.verbose, log_probs=log_probs
                 )
             yield chunk
+
+    def _get_max_tokens(self, prompt: str) -> int:
```
Comment (dev2049) [inline langchain/llms/llamacpp.py:291]: could we add a unit test for this? will probably want to make this a standalone function that takes a Llama client as a arg

File: langchain/llms/llamacpp.py [hunk 4]
```diff
@@ -230,6 +230,10 @@ def _call(
             return combined_text_output
         else:
             params = self._get_parameters(stop)
+
+            if params["max_tokens"] < 0:
```
Comment (dev2049) [inline langchain/llms/llamacpp.py:234]: should handle case where max_tokens is None

Comment (dev2049) [reply langchain/llms/llamacpp.py:234]: speaking of which, how does llama_cpp handle Nones for max_tokens? could it already be doing this inference under the hood?

**Comment by leo-gan** [comment]
@masands Hi , could you, please, resolve the merging issues and address the last comments (if needed)? After that, ping me and I push this PR for the review. Thanks!

**Comment by efriis** [comment]
Closing because the PR wouldn't line up with the current directory structure of the library (would need to be in /libs/langchain/langchain instead of /langchain). Feel free to reopen against the current head if it's still relevant!

**Pull request closed without merging**